extensions/csv/src/main/java/io/deephaven/csv/reading/CellGrabber.java

```diff
@@ -0,0 +1,276 @@
+        this.size = 0;
+        this.offset = 0;
+        this.startOffset = 0;
+        this.spillBuffer = new GrowableCharBuffer();
+    }
+
+    /**
+     * Try to grab the next cell from the input, being aware of field delimiters, line delimiters,
+     * quoting, and trimming.
+     * @param dest The result, as a {@link CharSlice}. The CharSlice is invalidated by the next
+     *             call to {@link #grabNext}.
+     * @param lastInRow An out parameter. lastInRow[0] will be set to true iff the cell just read
+     *                  was the last cell in the row.
+     * @return true if a cell was read; false if at end of input.
+     */
+    public boolean grabNext(final CharSlice dest, final boolean[] lastInRow) throws CsvReaderException {
+        spillBuffer.clear();
+        startOffset = offset;
+
+        if (ignoreSurroundingSpaces) {
+            skipWhitespace();
+        }
+        if (!tryEnsureMore()) {
+            return false;
+        }
+
+        // Is first char the quote char?
+        if (buffer[offset] == quoteChar) {
+            ++offset;
+            processQuotedMode(dest, lastInRow);
+            if (trim) {
+                trimWhitespace(dest);
+            }
+        } else {
+            processUnquotedMode(dest, lastInRow);
+        }
+        return true;
+    }
+
+    private void processQuotedMode(final CharSlice dest, boolean[] lastInRow) throws CsvReaderException {
+        startOffset = offset;
+        OUTER: while (true) {
+            while (offset != size) {
+                final char ch = buffer[offset++];
+                if (ch != quoteChar) {
+                    // Ordinary character. Note: in quoted mode we will gladly eat field and line separators.
+                    continue;
+                }
+                // This character is a quote char. It could be the end of the cell, or it could be an escaped
+                // quote char (e.g. ""). The way to tell is to peek ahead at the next character.
+                if (!tryEnsureMore()) {
+                    // There is no next char (we are at end of input), so let's call this end of cell.
+                    break OUTER;
+                }
+                final char peek = buffer[offset];
+                if (peek != quoteChar) {
+                    // There is a next char, but it's not a quotation mark. So this
+                    // quotation mark must be the end of the quoted string.
+                    break OUTER;
+                }
+                // There is a next character, and it *is* a quotation mark. So this is a quoted quote
+                // "", to be interpreted as ". So we'll spill this string (up to the first quotation mark),
+                // skip the second quotation mark, and keep going.
+                spillRange();
+                // Skip the second quotation mark.
+                ++offset;
+                startOffset = offset;
+            }
+            if (!tryEnsureMore()) {
```

Comment:
Do we need two nested while loops here?  Sounds like we could combine the inner while condition as a second condition in the if check here and avoid nested whiles?  Haven't tried it, just suggesting you explore it, if the code ends up being more complicated/harder to read, drop.

Reply:
Nice catch. I think you're right. Changed.